mongo_connector/doc_managers/neo4j_doc_manager.py

```diff
@@ -53,16 +58,73 @@ def upsert(self, doc, namespace, timestamp):
         "ns": namespace,
         "_ts": timestamp
     }
+    doc = self._formatter.format_document(doc)
+    self.store_nodes_and_relationships(doc, doc_type, doc_id)
+    
   
+  def store_nodes_and_relationships(self, doc, doc_type, doc_id):
+    self.build_nodes(doc_type, doc, doc_id)
+    for node_type in self.doc_types:
+      self.apply_id_constraint(node_type)
+    tx = self.graph.cypher.begin()
+    for statement in self.query_nodes.keys():
+      tx.append(statement, {"parameters":self.query_nodes[statement]})
+    main_type = self.doc_types.pop(0)
+    for node_type in self.doc_types:
+      tx.append(self.build_relationships(main_type, node_type), {"doc_id": doc_id})
+    tx.commit()
+    self.query_nodes = {}
+    self.doc_types = []
+
+  def build_nodes(self, doc_type, hash, id):
+    self.doc_types.append(doc_type)
+    parameters = {'_id':id}
+    for key in hash.keys():
+      if (type(hash[key]) is dict):
+        self.build_nodes(key, hash[key], id)
+      elif ((type(hash[key]) is list) and (type(hash[key][0]) is dict)):
+        for json in hash[key]:
+          json_key = key + str(hash[key].index(json))
+          self.build_nodes(json_key, json, id)
+      else:
+        value = hash[key]
+        parameters.update({ key: value })
+    query = "CREATE (c:Document:{doc_type} {{parameters}})".format(doc_type=doc_type)
+    self.query_nodes.update({query: parameters})
+
+  def build_relationships(self, main_type, node_type):
+    relationship_type = main_type + "_" + node_type
+    statement = "MATCH (a:{main_type}), (b:{node_type}) WHERE a._id={{doc_id}} AND b._id ={{doc_id}} CREATE (a)-[r:{relationship_type}]->(b)".format(main_type=main_type, node_type=node_type, relationship_type=relationship_type)
+    return statement
+
   def bulk_upsert(self, docs, namespace, timestamp):
     """Insert multiple documents into Neo4j."""
     LOG.error("Bulk")
 
   def update(self, document_id, update_spec, namespace, timestamp):
-    LOG.error("Update")
+    self.commit()
+    doc_id = u(document_id)
+    update_value_list = update_spec['$set']
+    index, doc_type = self._index_and_mapping(namespace)
+    tx = self.graph.cypher.begin()
+    params_dict = {"doc_id": doc_id}
+    set_dict = {}
+    for update_value in update_value_list.keys():
+      set_dict.update({update_value: update_value_list[update_value]})
+    params_dict.update({"set_parameter": set_dict})
+    statement = "MATCH (d:Document:{doc_type}) WHERE d._id={{doc_id}} SET d+={{set_parameter}}".format(doc_type=doc_type)
+    tx.append(statement, params_dict)
+    tx.commit()
 
   def remove(self, document_id, namespace, timestamp):
-    LOG.error("remove")
+    """Removes a document from Neo4j."""
+    doc_id = u(document_id)
+    index, doc_type = self._index_and_mapping(namespace)
+    params_dict = {"doc_id": doc_id}
+    tx = self.graph.cypher.begin()
+    statement = "MATCH (d:Document) WHERE d._id={doc_id} OPTIONAL MATCH (d)-[r]-() DELETE d, r"
```

Comment:
Still in a large graph this operation would then be O(n) which we definitely can't afford, also those deletes will block other operations.

You don't have to add it to this PR. We just must not forget about it.